### ISSUE TITLE
[BUG] Running out of PG Connections

### ISSUE THREAD

**Issue by slimdevl** [Permify/permify#768]
**Describe the bug**
I am not sure this is a bug, but maybe something I am doing. I have the service up in `docker-container`, and have a relatively simple model. (basically the `user groups` example): https://play.permify.co/?s=user-groups&tab=schema

I am running tests which add, then delete a series of relationships. I am doing this across a number of threads in my test.
1. add user to org as admin
2. add user to org as member 
3. add organization to team as org
4. add user to team owner
5. add user to team member
6. add organization to project as team
7. add team to project as team
8. then reverse... `7, 6, 5, 4, 3, 2, 1`

Every so often it fails, and the logs have the following:
```json
{"level":"info","time":"2023-10-24T00:44:44Z","caller":"/home/runner/work/permify/permify/pkg/cmd/serve.go:305","message":"🚀 http server successfully started: 3476"}
{"level":"info","time":"2023-10-24T00:48:14Z","caller":"/home/runner/work/permify/permify/internal/storage/postgres/utils/common.go:118","message":"failed to rollback transaction%!(EXTRA *pgconn.connLockError=conn closed)"}
{"level":"info","time":"2023-10-24T00:48:15Z","caller":"/home/runner/work/permify/permify/internal/storage/postgres/utils/common.go:118","message":"failed to rollback transaction%!(EXTRA *pgconn.errTimeout=timeout: context already done: context canceled)"}
{"level":"info","time":"2023-10-24T00:48:15Z","caller":"/home/runner/work/permify/permify/internal/storage/postgres/utils/common.go:118","message":"failed to rollback transaction%!(EXTRA *pgconn.errTimeout=timeout: context already done: context canceled)"}
{"level":"info","time":"2023-10-24T00:48:16Z","caller":"/home/runner/work/permify/permify/internal/storage/postgres/utils/common.go:118","message":"failed to rollback transaction%!(EXTRA *pgconn.errTimeout=timeout: context already done: context canceled)"}
{"level":"info","time":"2023-10-24T00:48:16Z","caller":"/home/runner/work/permify/permify/internal/storage/postgres/utils/common.go:118","message":"failed to rollback transaction%!(EXTRA *pgconn.errTimeout=timeout: context already done: context canceled)"}
{"level":"info","time":"2023-10-24T00:48:26Z","caller":"/home/runner/work/permify/permify/internal/storage/postgres/utils/common.go:118","message":"failed to rollback transaction%!(EXTRA *pgconn.errTimeout=timeout: context already done: context canceled)"}
```

I am hoping to get more insight into these - and why they might be happening? I figure I am doing something wrong, or maybe pushing the system to hard, and running out of `pg connections`. Is there any more flags I can turn on to get more data on the pools inside the service?

**To Reproduce**
using https://play.permify.co/?s=user-groups&tab=schema

In parallel threads, do the following relationship adds, then deletes:
1. add user to org as admin
2. add user to org as member 
3. add organization to team as org
4. add user to team owner
5. add user to team member
6. add organization to project as team
7. add team to project as team
8. then reverse... `7, 6, 5, 4, 3, 2, 1`

Here is a sample of my code and/or tests that demonstrate the issue in my app:

**Expected behavior**
Things just work - but that is coming ;-)

**Additional context**

**Environment (please complete the following information, because it helps us investigate better):**
- Linux Ubuntu 22.04 
- Docker Compose Service + Postgres Local
- `go test` running parallel cases of the same thing :-)
- `golang client for making calls to service`

**Comment by slimdevl** [comment]
> Hello @slimdevl , which version of Permify and which version of Postgres are you using?

0.5.4

**Comment by tolgaOzen** [comment]
Hi @slimdevl , did the migrations run? Can you see the tables in your database?

**Comment by slimdevl** [comment]
Sorry - have been getting crushed at work. If you look at this: https://github.com/slimdevl/permify-perf-repo - which is really close how the system I am building works, you will be able to put some pressure on the system. 

I believe what I was seeing earlier was related two things:
1. I didn't have the rate limiter in my client. 
2. I would run once (say 1 by 100) and the second run would generate the errors above.

That said, I can still make some errors occur with multiple runs (under the rate-limit) using the reference test setup.

Please check it out - would be interested in your feedback. 

BTW - thanks for your help :-) This project is awesome

**Comment by tolgaOzen** [comment]
Hello @slimdevl , we added a backoff strategy in version 0.7.1 to overcome the issue with the writes. I ran the test you wrote and didn't encounter any errors. Could you try it again on your end? If you also don't receive any errors, we can close the issue

**Comment by EgeAytin** [comment]
Closing this issue, seems like no problem exists.
